Macrotracker support

Duhemm opened the pull request:
Hi !

> MacroTracker is a compiler plugin for scalac 2.11.0. Its job is to register all the symbols that are touched during the expansion of a macro, and attach them to the expanded macro. Using this information, we can deduct the dependencies between the expanded macro and everything that has been involved in its expansion. 

(From [this thread in sbt-dev](https://groups.google.com/forum/#!topic/sbt-dev/clIbzCp9VX8))

MacroTracker helps to track the dependencies that are introduced using the reflection API. Whenever a symbol is accessed using the reflection API, MacroTracker registers it and finally attaches all the touched symbols to the expanded macro.

It can help greatly in cases such as the ones that are exposed in the [accompanied scripted test](https://github.com/Duhemm/sbt/tree/macrotracker-support/sbt/src/sbt-test/source-dependencies/macro-macrotracker).

MacroTracker can be found here : [scalamacros/macrotracker](https://github.com/scalamacros/macrotracker).

Commit:
Implement support for MacroTracker in sbt

MacroTracker is a scala compiler plugin that attach to expanded macros
all the symbols that have been touched using the reflection API.

sbt can then use this information to reason about the dependencies of
the macro client, and will recompile the macro client whenever a file
that is inspected using reflection is modified.

Commit:
Add a scripted test for macrotracker

typesafe-tools commented:
Can one of the admins verify this patch?

xeno-by commented:
Looks like the test failure is spurious. Would it be possible to rerun the test suite? 

Also, wdyt about this patch? From what I can see, it allows achieving much better precision than what was possible before + brings us closer to closing https://github.com/sbt/sbt/issues/1729.

eed3si9n commented:
I'm rerunning the tests.

eed3si9n commented:
@jsuereth and I were both busy during December with Typesafe related things, but hopefully we can review open PRs including this one after we come back from the break next week.

Commit:
No transitive invalidation with Macrotracker

The previous implementation performed a transitive invalidation of all
files that depend by macro expansion on a file whose API is modified.

This behavior was not correct and has been corrected so that only the
files that directly depend by macro expansion on a file whose API is
modified are invalidated.

xeno-by commented:
Hello guys! Any news on this one?

gkossakowski commented:
I have looked at this PR from very high-level point of view. It's great to see that work done in #1340 is paying off: major feature is being added without too many changes to incremental compiler sources.

I have a few concerns:
- this PR introduces a dependency on a compiler plugin that is of SNAPSHOT version. SNAPSHOTs are not guaranteed to stay around but sbt tests should run for a long time. We'll need a stable release at the very least.
- touched symbols are stored as tree attachment at the moment. Is this how you guys this envision long term?
- we run tree traverser that looks for attached symbols on all tree nodes in every file, no matter if there's any macro expansion or not; what kind of performance hit are we expecting?

Also, do you see your compiler plugin merged back into Scala 2.11/2.12? It would really help to understand long-term plans around this feature.

xeno-by commented:
1) I can push a stable release if necessary.

2) I think so, yes. This is domain-specific metadata (much like e.g. `MacroExpansionAttachment`), so I don't have other ideas how to associate it with trees except using attachments. Do you see any problems that might result from this?

3) I didn't give any thought to backporting this functionality into scala/scala. If you find it useful and if the Scala team finds it reasonable, I can submit a pull request that embeds tracking of touchedSymbols into Scala 2.11. /cc @retronym

retronym commented:
I would favour a minmial change to SBT that enabled the meat of this PR to be factored out as a SBT plugin. But I fear that this will prove a difficult task.

Failing that, we need another way to be sure that the new code is only run when users opt-in to the macro tracking compiler plugin. Perhaps you can check whether that plugin is enabled. We don't want other users exposed to performance penalties of extra traversals, nor to potential bugs in this code.

I took a look at the internals of macro-tracker. It looks to be rather difficult to maintain. How can we reason that all overrides are provided? I noticed that the inherent difficulty, and the solution, runtime reflection synchronization. I suppose the macro tracker has a lower bar, we shouldn't reject a solution that 99% of dependencies just because we can't prove it tracks 100%. But I would love to hear ideas for how we could implement this more cleanly if it were done in the compiler proper.

gkossakowski commented:
Making incremental fully extensible might be rather challenging. It might be more work than providing support for a few special features.
I think structuring the code so this logic is clearly separated and making the whole feature opt-in is a good approach.
I'd suggest to add logging of timings so users can easily understand the performance hit they are opting-in.

xeno-by commented:
@retronym The only approach that I can think of involves macros, which might or might not be desireable in the compiler (like the experience with runtime reflection has shown).

`Perhaps you can check whether that plugin is enabled`
I find this be a really great idea!

`performance penalties of extra traversals`
Can we fuse this traversal into the other ones that are already performed?

`if it were done in the compiler proper`
Does this mean that you'd be interested in having something like macrotracker in scala/scala?

retronym commented:
> Does this mean that you'd be interested in having something like macrotracker in scala/scala?

Take it as meaning: we could start talking about that if implementation in the compiler is drastically simpler than the implementation as a plugin :)

xeno-by commented:
@retronym I could write a macro that generates a proxy that mechanically calls wrap/unwrap and lets compiler developers to provide custom implementations of wrap/unwrap for the types they are interested in. Would that qualify as being drastically simpler?

retronym commented:
The macro itself will still take some work to understand. And we don't have great tools for debugging code involving macros, so I'd be a little wary about that.

But it does seem like a worthwhile thing to try if you have the time. I guess the refactoring to use a macro rather than boilerplate could be done in the compiler plugin as in the core compiler, though.

dwijnand commented:
Tracked in https://github.com/sbt/zinc/issues/249.
